Mathlib/Tactic/Congr!.lean

```diff
@@ -0,0 +1,248 @@
+If successful, returns the new goal, and otherwise returns the original `MVarId`.
+-/
+def Lean.MVarId.iffOfEq (mvarId : MVarId) : MetaM MVarId := do
+  let res ← observing? do
+    let [mvarId] ← mvarId.apply (mkConst `iff_of_eq []) | failure
+    return mvarId
+  return res.getD mvarId
+
+/--
+Try to convert an `Eq` into an `Iff` by applying `propext`.
+If successful, then returns then new goal, otherwise returns the original `MVarId`.
+-/
+def Lean.MVarId.propext (mvarId : MVarId) : MetaM MVarId := do
+  let res ← observing? do
+    let [mvarId] ← mvarId.apply (mkConst `propext []) | failure
+    return mvarId
+  return res.getD mvarId
+
+/-- Helper theorem for `LEan.MVar.liftReflToEq`. -/
+theorem Lean.MVarId.rel_of_eq_and_refl {R : α → α → Prop} (hxy : x = y) (h : R x x) :
+    R x y := hxy ▸ h
+
+/--
+Use a `refl`-tagged lemma to convert the goal into an `Eq`. If this can't be done, returns
+the original `MVarId`.
+-/
+def Lean.MVarId.liftReflToEq (mvarId : MVarId) : MetaM MVarId := do
+  mvarId.checkNotAssigned `liftReflToEq
+  let tgt ← withReducible <| mvarId.getType'
+  let .app (.app rel _) _ := tgt | return mvarId
+  if rel.isAppOf `Eq then
+    -- No need to lift Eq to Eq
+    return mvarId
+  let reflLemmas ← (Mathlib.Tactic.reflExt.getState (← getEnv)).getMatch rel
+  for lem in reflLemmas do
+    let res ← observing? do
+      -- First create an equality relating the LHS and RHS
+      -- and reduce the goal to proving that LHS is related to LHS.
+      let [mvarIdEq, mvarIdR] ←
+            mvarId.apply (← mkConstWithFreshMVarLevels ``Lean.MVarId.rel_of_eq_and_refl)
+        | failure
+      -- Then fill in the proof of the latter by reflexivity.
+      let [] ← mvarIdR.apply (← mkConstWithFreshMVarLevels lem) | failure
+      return mvarIdEq
+    if let some mvarId := res then
+      return mvarId
+  return mvarId
+
+/--
+Try to close the goal using `Subsingleton.elim`. Returns whether or not it succeeds.
+-/
+def Lean.MVarId.subsingletonElim (mvarId : MVarId) : MetaM Bool := do
+  let res ← observing? do
+    let [] ← mvarId.apply (mkConst ``Subsingleton.elim [← mkFreshLevelMVar]) | failure
+    return true
+  return res.getD false
+
+def Lean.MVarId.proofIrrelHeq (mvarId : MVarId) : MetaM Bool := do
+  let res ← observing? do
+    let [] ← mvarId.apply (mkConst `proof_irrel_heq []) | failure
+    return true
+  return res.getD false
+
+/--
+Try to apply `pi_congr`. This is similar to `Lean.MVar.congrImplies?`.
+-/
+def Lean.MVarId.congrPi? (mvarId : MVarId) : MetaM (Option (List MVarId)) :=
+  observing? do mvarId.apply (← mkConstWithFreshMVarLevels `pi_congr)
+
+/--
+Try to apply `funext`, but only if it is an equality of two functions where at least one is
+a lambda expression.
+One thing this check prevents is accidentally applying `funext` to a set equality, but also when
+doing congruence we don't want to apply `funext` unnecessarily.
+"Obvious" means that the type of the terms being equated is a pi type.
```

Comment:
Not sure what you mean here? You don't check this, only that one of the terms is a lambda-expression, right?

Reply:
Oops, that was a remnant of an earlier version. (It was not a very good idea.)